src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -140,22 +140,28 @@ def and_constraints(parsed_constraints: Sequence[Requirement]) -> Requirement:
             merged_specs: Set[Tuple[str, str]] = set()
             expected_interpreter = parsed_constraints[0].project_name
             for parsed_constraint in parsed_constraints:
-                if parsed_constraint.project_name != expected_interpreter:
-                    attempted_interpreters = {
-                        interp: sorted(
-                            str(parsed_constraint) for parsed_constraint in parsed_constraints
-                        )
-                        for interp, parsed_constraints in itertools.groupby(
-                            parsed_constraints,
-                            key=lambda pc: pc.project_name,
-                        )
-                    }
-                    raise ValueError(
-                        "Tried ANDing Python interpreter constraints with different interpreter "
-                        "types. Please use only one interpreter type. Got "
-                        f"{attempted_interpreters}."
+                if parsed_constraint.project_name == expected_interpreter:
+                    merged_specs.update(parsed_constraint.specs)
+                    continue
+
+                def key_fn(req: Requirement):
```

Comment:
needs a return type?

Reply:
I think it's fine without, as this is an inner function and the return type of a key function doesn't matter much. What we had before was a `lambda`, which has zero typing.